File: SOLID/OpenClosePrinciple/ItinerarySearchEngine.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace SOLID.OpenClosePrinciple
{

    public class ItinerarySearchEngine
    {

        public Itinerary OptimalItinerary(Trip trip, ItineraryPreference pref)
        {
            return pref switch
            {
                ItineraryPreference.Cheapest => ItinerariesFor(trip).OrderBy(i => i.Cost).FirstOrDefault(),
                ItineraryPreference.Shortest => ItinerariesFor(trip).OrderBy(i => i.Duration).FirstOrDefault(),
                _ => throw new ArgumentException("Unknown ItineraryType: " + pref)
            };
        }

        private static IEnumerable<Itinerary> ItinerariesFor(Trip trip)
        {
            //fake implementation
            Itinerary directFlight = Itinerary.Of(trip)
                .Labelled("Direct flight")
                .Lasting(TimeSpan.FromHours(12))
                .Connecting(0)
                .Costing(400);

            Itinerary withDubaiStopover = Itinerary.Of(trip)
                .Labelled("With Dubai stopover")
                .Lasting(TimeSpan.FromHours(16))
                .Connecting(1)
                .Costing(200);

            return new[] { directFlight, withDubaiStopover };
        }

    }
}

File: Java/Exercices/workspaces/solid-ocp-start/src/main/java/zenika/solid/ocp/ItinerarySearchEngine.java
package zenika.solid.ocp;

import java.time.Duration;
import java.util.Optional;
import java.util.stream.Stream;

import static java.util.Comparator.comparing;

public class ItinerarySearchEngine {
    
    public Optional<Itinerary> optimalItinerary(Trip trip, ItineraryPreference pref) {
        switch(pref) {
            case CHEAPEST: 
                return itinerariesFor(trip).min(comparing(Itinerary::cost));
            case SHORTEST: 
                return itinerariesFor(trip).min(comparing(Itinerary::duration));
            default:
                throw new IllegalArgumentException("Unknown ItineraryType: " + pref);
        }
    }

    private Stream<Itinerary> itinerariesFor(Trip trip) {
        //fake implementation
        Itinerary directFlight = Itinerary.of(trip)
                .labelled("Direct flight")
                .lasting(Duration.ofHours(12))
                .costing(400);
        
        Itinerary withDubaiStopover = Itinerary.of(trip)
                .labelled("With Dubai stopover")
                .lasting(Duration.ofHours(16))
                .costing(200);
        
        return Stream.of(directFlight, withDubaiStopover);
    }

}
